boskos/metrics/resources.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/boskos/common"
+	"k8s.io/test-infra/boskos/ranch"
+)
+
+const (
+	// ResourcesMetricName is the name of the Prometheus metric used to monitor Boskos resources.
+	ResourcesMetricName = "boskos_resourcess"
+	// ResourcesMetricDescription is the description for the Prometheus metric used to monitor Boskos resources.
+	ResourcesMetricDescription = "Number of resources recorded in Boskos by resource type and state."
+)
+
+var (
+	// ResourcesMetricLabels is the list of labels used for the Prometheus metric used to monitor Boskos resources.
+	ResourcesMetricLabels = []string{"type", "state"}
+)
+
+type resourcesCollector struct {
+	boskosResources *prometheus.Desc
+	ranch           *ranch.Ranch
+}
+
+// NewResourcesCollector returns a collector which exports the current counts of
+// Boskos resources, segmented by resource type and state.
+func NewResourcesCollector(ranch *ranch.Ranch) prometheus.Collector {
+	return resourcesCollector{
+		boskosResources: prometheus.NewDesc(ResourcesMetricName, ResourcesMetricDescription, ResourcesMetricLabels, nil),
+		ranch:           ranch,
+	}
+}
+
+func (rc resourcesCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- rc.boskosResources
+}
+
+func (rc resourcesCollector) Collect(ch chan<- prometheus.Metric) {
+	metrics, err := rc.ranch.AllMetrics()
+	if err != nil {
+		logrus.WithError(err).Error("failed to get metrics")
+	}
+	NormalizeResourceMetrics(metrics, common.KnownStates, func(rtype, state string, count float64) {
+		ch <- prometheus.MustNewConstMetric(rc.boskosResources, prometheus.GaugeValue, count, rtype, state)
+	})
+}
+
+// NormalizeResourceMetrics "normalizes" the list of provided Metrics by
+// bucketing any state not in states into the "Other" state, and by ensuring
+// every state in states has some count (even if zero).
+// It then applies the function for each combination of resource type and state.
+func NormalizeResourceMetrics(metrics []common.Metric, states []string, updateFunc func(rtype, state string, count float64)) {
```

Comment:
I'm open to a better name than "Normalize".

Reply:
You don't need to keep that functionality, that can probably be done in the dashboard. This was done for mason resources, since they have a name of a different resource as a state. Up to you.

Reply:
The [Prometheus docs do warn against having too many different label values](https://prometheus.io/docs/practices/naming/#labels), so I think it makes sense to have a catch-all "other" category.
> CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.